Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,430 @@
+
+            if (IsInvokeCommandSessionScriptBlock(cmdName, commandAst))
+            {
+                var sessionName = GetSessionName(commandAst);
+
+                var varsInLocalAssignments = FindVarsInAssignmentAsts(scriptBlockExpressionAst);
+                if (varsInLocalAssignments != null)
+                {
+                    AddAssignedVarsToSession(sessionName, varsInLocalAssignments);
+                }
+
+                GenerateDiagnosticRecords(
+                    FindNonAssignedNonUsingVarAsts(
+                        scriptBlockExpressionAst, 
+                        GetAssignedVarsInSession(sessionName)));
+
+                return AstVisitAction.SkipChildren;
+            }
+            
+            return AstVisitAction.Continue;
+        }
+
+        /// <summary>
+        /// FindVarsInAssignmentAsts: Retrieves all assigned variables from an Ast:
+        /// Example: `$foo = "foo"` ==> the VariableExpressionAst for $foo is returned
+        /// </summary>
+        /// <param name="ast"></param>
+        private static IEnumerable<VariableExpressionAst> FindVarsInAssignmentAsts(Ast ast)
+        {
+            // Find all variables that are assigned within this ast
+            return ast.FindAll(
+                    predicate: a => a is VariableExpressionAst varExpr &&
+                                    varExpr.Parent is AssignmentStatementAst assignment &&
+                                    assignment
+                                        .Left
+                                        .Equals(varExpr),
+                    searchNestedScriptBlocks: true)
+                .Select(a => a as VariableExpressionAst);
+        }
+
+        /// <summary>
+        /// FindNonAssignedNonUsingVarAsts: Retrieve variables that are:
+        /// - not assigned before
+        /// - not prefixed with the 'Using' scope modifier
+        /// - not a PowerShell special variable
+        /// </summary>
+        /// <param name="ast"></param>
+        /// <param name="varsInAssignments"></param>
+        private static List<VariableExpressionAst> FindNonAssignedNonUsingVarAsts(
+            Ast ast, IEnumerable<VariableExpressionAst> varsInAssignments)
+        {
+            // Find all variables that are not locally assigned, and don't have $using: scope modifier
+            return ast.FindAll(
+                    predicate: a => a is VariableExpressionAst varAst &&
+                                    !(varAst.Parent is UsingExpressionAst) &&
+                                    varsInAssignments.All(
+                                        b => b.VariablePath.UserPath != varAst.VariablePath.UserPath) &&
+                                    !Helper
+                                        .Instance
+                                        .HasSpecialVars(varAst.VariablePath.UserPath),
+                    searchNestedScriptBlocks: true)
+                .Select(a => a as VariableExpressionAst).ToList();

```

Comment:
Here you should definitely use a `foreach` and build the list directly to prevent double allocation